In [1]:
import pandas as pd

## Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [2]:
horse_data = pd.read_csv('horse_data.csv', na_values = '?', names = ['surgery?', 'Age', 'Hospital Number', 'rectal temperature', 
                                                    'pulse', 'respiratory rate', 'temperature of extremities', 
                                                    'peripheral pulse', 'mucous membranes', 'capillary refill time', 
                                                    'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube', 
                                                    'nasogastric reflux', 'nasogastric reflux PH', 
                                                    'rectal examination - feces', 'abdomen', 'packed cell volume', 
                                                    'total protein', 'abdominocentesis appearance', 
                                                    'abdomcentesis total protein', 'outcome', 'surgical lesion?', 
                                                    'type of lesion_1', 'type of lesion_2', 'type of lesion_3', 'cp_data'])
horse_data

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0         2.0    1           530101                38.5   66.0   
1         1.0    1           534817                39.2   88.0   
2         2.0    1           530334                38.3   40.0   
3         1.0    9          5290409                39.1  164.0   
4         2.0    1           530255                37.3  104.0   
..        ...  ...              ...                 ...    ...   
295       1.0    1           533886                 NaN  120.0   
296       2.0    1           527702                37.2   72.0   
297       1.0    1           529386                37.5   72.0   
298       1.0    1           530612                36.5  100.0   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
0                28.0                         3.0               3.0   
1                20.0                         NaN               NaN   
2                24.0                         1.0               1.0   
3                84.0                         4.0               1.0   
4                35.0                         NaN               NaN   
..                ...                         ...               ...   
295              70.0                         4.0               NaN   
296              24.0                         3.0               2.0   
297              30.0                         4.0               3.0   
298              24.0                         3.0               3.0   
299              20.0                         NaN               NaN   

     mucous membranes  capillary refill time  ...  packed cell volume  \
0                 NaN                    2.0  ...                45.0   
1                 4.0                    1.0  ...                50.0   
2                 3.0                    1.0  ...                33.0   
3                 6.0                    2.0  ...                48.0   
4                 6.0                    2.0  ...                74.0   
..                ...                    ...  ...                 ...   
295               4.0                    2.0  ...                55.0   
296               4.0                    2.0  ...                44.0   
297               4.0                    1.0  ...                60.0   
298               3.0                    1.0  ...                50.0   
299               NaN                    NaN  ...                36.0   

     total protein  abdominocentesis appearance  abdomcentesis total protein  \
0              8.4                          NaN                          NaN   
1             85.0                          2.0                          2.0   
2              6.7                          NaN                          NaN   
3              7.2                          3.0                          5.3   
4              7.4                          NaN                          NaN   
..             ...                          ...                          ...   
295           65.0                          NaN                          NaN   
296            NaN                          3.0                          3.3   
297            6.8                          NaN                          NaN   
298            6.0                          3.0                          3.4   
299           62.0                          1.0                          1.0   

     outcome  surgical lesion?  type of lesion_1  type of lesion_2  \
0        2.0                 2             11300                 0   
1        3.0                 2              2208                 0   
2        1.0                 2                 0                 0   
3        2.0                 1              2208                 0   
4        2.0                 2              4300                 0   
..       ...               ...               ...               ...   
295      3.0                 2 

In [3]:
horse_data.describe()

surgery?         Age  Hospital Number  rectal temperature  \
count  299.000000  300.000000     3.000000e+02          240.000000   
mean     1.397993    1.640000     1.085889e+06           38.167917   
std      0.490305    2.173972     1.529801e+06            0.732289   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.800000   
50%      1.000000    1.000000     5.303055e+05           38.200000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory rate  temperature of extremities  \
count  276.000000        242.000000                  244.000000   
mean    71.913043         30.417355                    2.348361   
std     28.630557         17.642231                    1.045054   
min     30.000000          8.000000                    1.000000   
25%     48.000000         18.500000                    1.000000   
50%     64.000000         24.500000                    3.000000   
75%     88.000000         36.000000                    3.000000   
max    184.000000         96.000000                    4.000000   

       peripheral pulse  mucous membranes  capillary refill time  ...  \
count        231.000000        253.000000             268.000000  ...   
mean           2.017316          2.853755               1.305970  ...   
std            1.042428          1.620294               0.477629  ...   
min            1.000000          1.000000               1.000000  ...   
25%            1.000000          1.000000               1.000000  ...   
50%            2.000000          3.000000               1.000000  ...   
75%            3.000000          4.000000               2.000000  ...   
max            4.000000          6.000000               3.000000  ...   

       packed cell volume  total protein  abdominocentesis appearance  \
count          271.000000     267.000000                   135.000000   
mean            46.295203      24.456929                     2.037037   
std             10.419335      27.475009                     0.804905   
min             23.000000       3.300000                     1.000000   
25%             38.000000       6.500000                     1.000000   
50%             45.000000       7.500000                     2.000000   
75%             52.000000      57.000000                     3.000000   
max             75.000000      89.000000                     3.000000   

       abdomcentesis total protein     outcome  surgical lesion?  \
count                   102.000000  299.000000        300.000000   
mean                      3.019608    1.551839          1.363333   
std                       1.968567    0.737187          0.481763   
min                       0.100000    1.000000          1.000000   
25%                       2.000000    1.000000          1.000000   
50%                       2.250000    1.000000          1.000000   
75%                       3.900000    2.000000          2.000000   
max                      10.100000    3.000000          2.000000   

       type of lesion_1  type of lesion_2  type of lesion_3     cp_data  
count        300.000000        300.000000        300.000000  300.000000  
mean        3657.880000         90.226667          7.363333    1.670000  
std         5399.513513        649.569234        127.536674    0.470998  
min            0.000000          0.000000          0.000000    1.000000  
25%         2111.750000          0.000000          0.000000    1.000000  
50%         2673.500000          0.000000          0.000000    2.000000  
75%         3209.000000          0.000000          0.000000    2.000000  
max        41110.000000       7111.000000       2209.000000    2.000000  

[8 rows x 28 columns]

In [4]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery?                     299 non-null    float64
 1   Age                          300 non-null    int64  
 2   Hospital Number              300 non-null    int64  
 3   rectal temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory rate             242 non-null    float64
 6   temperature of extremities   244 non-null    float64
 7   peripheral pulse             231 non-null    float64
 8   mucous membranes             253 non-null    float64
 9   capillary refill time        268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal distension         244 non-null    float64
 13  nasogastric tube    

In [59]:
horse_8col = horse_data[ ['surgery?', 'Age', 'rectal temperature', 'pulse', 'temperature of extremities', 'abdominal distension', 'packed cell volume', 'total protein'] ]
horse_8col

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
0         2.0    1                38.5   66.0                         3.0   
1         1.0    1                39.2   88.0                         NaN   
2         2.0    1                38.3   40.0                         1.0   
3         1.0    9                39.1  164.0                         4.0   
4         2.0    1                37.3  104.0                         NaN   
..        ...  ...                 ...    ...                         ...   
295       1.0    1                 NaN  120.0                         4.0   
296       2.0    1                37.2   72.0                         3.0   
297       1.0    1                37.5   72.0                         4.0   
298       1.0    1                36.5  100.0                         3.0   
299       1.0    1                37.2   40.0                         NaN   

     abdominal distension  packed cell volume  total protein  
0                     4.0                45.0            8.4  
1                     2.0                50.0           85.0  
2                     1.0                33.0            6.7  
3                     4.0                48.0            7.2  
4                     NaN                74.0            7.4  
..                    ...                 ...            ...  
295                   NaN                55.0           65.0  
296                   3.0                44.0            NaN  
297                   3.0                60.0            6.8  
298                   3.0                50.0            6.0  
299                   NaN                36.0           62.0  

[300 rows x 8 columns]

In [6]:
horse_8col.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   temperature of extremities  244 non-null    float64
 5   abdominal distension        244 non-null    float64
 6   packed cell volume          271 non-null    float64
 7   total protein               267 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [7]:
# расчет моды
horse_8col_mode = horse_8col.mode()
horse_8col_mode.index = [(f'mode_{x}') for x in list(horse_8col_mode.index)]
horse_8col_mode

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
mode_0       1.0  1.0                38.0   48.0                         3.0   
mode_1       NaN  NaN                 NaN    NaN                         NaN   

        abdominal distension  packed cell volume  total protein  
mode_0                   1.0                37.0            6.5  
mode_1                   NaN                50.0            7.0

In [102]:
# расчитаем базовые метрики
stat_data = horse_8col.describe()
stat_data

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       temperature of extremities  abdominal distension  packed cell volume  \
count                  244.000000            244.000000          271.000000   
mean                     2.348361              2.266393           46.295203   
std                      1.045054              1.065131           10.419335   
min                      1.000000              1.000000           23.000000   
25%                      1.000000              1.000000           38.000000   
50%                      3.000000              2.000000           45.000000   
75%                      3.000000              3.000000           52.000000   
max                      4.000000              4.000000           75.000000   

       total protein  
count     267.000000  
mean       24.456929  
std        27.475009  
min         3.300000  
25%         6.500000  
50%         7.500000  
75%        57.000000  
max        89.000000

In [99]:
# размах
x= {}
for name_col in horse_8col.columns:
    x[name_col] = [(stat_data[name_col]['max'] - stat_data[name_col]['min'])]
# x
scope = pd.DataFrame.from_dict(x)
scope.index = ['scope']

scope

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
scope       1.0  8.0                 5.4  154.0                         3.0   

       abdominal distension  packed cell volume  total protein  
scope                   3.0                52.0           85.7

In [107]:
# соединим в один dataframe рассчитанные выше показатели
basic_stat = pd.concat([stat_data, horse_8col_mode, scope])
basic_stat

surgery?         Age  rectal temperature       pulse  \
count   299.000000  300.000000          240.000000  276.000000   
mean      1.397993    1.640000           38.167917   71.913043   
std       0.490305    2.173972            0.732289   28.630557   
min       1.000000    1.000000           35.400000   30.000000   
25%       1.000000    1.000000           37.800000   48.000000   
50%       1.000000    1.000000           38.200000   64.000000   
75%       2.000000    1.000000           38.500000   88.000000   
max       2.000000    9.000000           40.800000  184.000000   
mode_0    1.000000    1.000000           38.000000   48.000000   
mode_1         NaN         NaN                 NaN         NaN   
scope     1.000000    8.000000            5.400000  154.000000   

        temperature of extremities  abdominal distension  packed cell volume  \
count                   244.000000            244.000000          271.000000   
mean                      2.348361              2.266393           46.295203   
std                       1.045054              1.065131           10.419335   
min                       1.000000              1.000000           23.000000   
25%                       1.000000              1.000000           38.000000   
50%                       3.000000              2.000000           45.000000   
75%                       3.000000              3.000000           52.000000   
max                       4.000000              4.000000           75.000000   
mode_0                    3.000000              1.000000           37.000000   
mode_1                         NaN                   NaN           50.000000   
scope                     3.000000              3.000000           52.000000   

        total protein  
count      267.000000  
mean        24.456929  
std         27.475009  
min          3.300000  
25%          6.500000  
50%          7.500000  
75%         57.000000  
max         89.000000  
mode_0       6.500000  
mode_1       7.000000  
scope       85.700000

Наибольшее отклонение наблюдается у следующих показателей: rectal temperature, packed cell volume и total protein. У этих же показателей наибольший разброс.
По показателям packed cell volume и total protein в моду попадает по два значения.
Показатель Age по всем квартилям равняется 1, а max у него 9. Это скорее всего выброс.

## Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [195]:
horse_8col

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
0         2.0    1                38.5   66.0                         3.0   
1         1.0    1                39.2   88.0                         NaN   
2         2.0    1                38.3   40.0                         1.0   
3         1.0    9                39.1  164.0                         4.0   
4         2.0    1                37.3  104.0                         NaN   
..        ...  ...                 ...    ...                         ...   
295       1.0    1                 NaN  120.0                         4.0   
296       2.0    1                37.2   72.0                         3.0   
297       1.0    1                37.5   72.0                         4.0   
298       1.0    1                36.5  100.0                         3.0   
299       1.0    1                37.2   40.0                         NaN   

     abdominal distension  packed cell volume  total protein  
0                     4.0                45.0            8.4  
1                     2.0                50.0           85.0  
2                     1.0                33.0            6.7  
3                     4.0                48.0            7.2  
4                     NaN                74.0            7.4  
..                    ...                 ...            ...  
295                   NaN                55.0           65.0  
296                   3.0                44.0            NaN  
297                   3.0                60.0            6.8  
298                   3.0                50.0            6.0  
299                   NaN                36.0           62.0  

[300 rows x 8 columns]

In [230]:
#  проверочная ячейка

q1 = horse_8col['total protein'].quantile(0.25)
q3 = horse_8col['total protein'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
print(lower_bound)
upper_bound = q3 + (1.5 * iqr)
print(upper_bound)

# remove_outliers = horse_8col[horse_8col['pulse'].between(lower_bound, upper_bound, inclusive=True)]

outliers_low = horse_8col[horse_8col['total protein'] < lower_bound]
outliers_up = horse_8col[horse_8col['total protein'] > upper_bound]
outliers_concat = pd.concat([outliers_low, outliers_up])

outliers_concat

-69.25
132.75


Empty DataFrame
Columns: [surgery?, Age, rectal temperature, pulse, temperature of extremities, abdominal distension, packed cell volume, total protein]
Index: []

In [21]:
def outliers(x, y, z):
    """ 
    Функция определяющая выбросы в DataFrame, где 1-ый аргумент - DataFrame, к которому применятеся функция, 
    2-ой аргумент - столбец, в котором осуществляется поиск выбросов,
    3-ий аргумент - изменяет результат работы функции: если поставить 0, то возвращаются ячейки с выбросами, 
    если 1, то возвращаются все ячейки, за исключением выбросов и ячеек, в которых в данном столбце стоит NaN. 
    
    """
    q1 = x[y].quantile(0.25)
    q3 = x[y].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)

# remove_outliers = horse_8col[horse_8col['pulse'].between(lower_bound, upper_bound, inclusive=True)]

    outliers_low = x[x[y] < lower_bound]
    outliers_up = x[x[y] > upper_bound]
    outliers_concat = pd.concat([outliers_low, outliers_up])
    
    remove_outliers = x[x[y].between(lower_bound, upper_bound, inclusive=True)]
    
    if z == 0:
        return outliers_concat
    else:
        return remove_outliers

In [23]:
?outliers

In [22]:
outliers(horse_8col, 'rectal temperature', 0)

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
44        1.0    1                35.4  140.0                         3.0   
80        1.0    1                36.4   98.0                         3.0   
118       1.0    1                36.5   78.0                         1.0   
141       2.0    1                36.0   42.0                         NaN   
238       2.0    1                36.1   88.0                         3.0   
251       2.0    1                36.6   42.0                         3.0   
298       1.0    1                36.5  100.0                         3.0   
20        1.0    1                39.9   72.0                         1.0   
54        2.0    1                40.3  114.0                         3.0   
75        1.0    9                39.7  100.0                         3.0   
91        2.0    1                40.3  114.0                         3.0   
99        2.0    1                39.6  108.0                         3.0   
259       1.0    1                40.8   72.0                         3.0   
281       2.0    1                40.0   78.0                         3.0   

     abdominal distension  packed cell volume  total protein  
44                    NaN                57.0           69.0  
80                    2.0                47.0            6.4  
118                   1.0                34.0           75.0  
141                   NaN                64.0            6.8  
238                   2.0                45.0            7.0  
251                   1.0                52.0            7.1  
298                   3.0                50.0            6.0  
20                    4.0                46.0            6.1  
54                    3.0                57.0            8.1  
75                    NaN                48.0           57.0  
91                    3.0                57.0            8.1  
99                    3.0                59.0            8.0  
259                   1.0                54.0            7.4  
281                   1.0                66.0            6.5

Нормальная температура лошади должна находиться в пределах 37.5-38.5 градусов Цельсия. Считается, что небольшое повышение температуры (от 38.6 до 39.4 градусов) вполне возможно после физической нагрузки лошади. В случае же, если температура тела лошади выше 39.5 градусов, есть подозрение на какое-то заболевание, и потому вам стоит обратиться за помощью к ветеринару. В редких случаях температура может быть ниже нормы (37.5 градусов). Такое возможно в случае длительного переохлаждения или если лошадь находится в состоянии шока. 

Источник: Важные моменты при осмотре заболевающей лошади | Уход за лошадью © fourhoofs.ru

Я считаю, что данные выбросы являются реальными показателями, т.к. незначительно отклоняются от нормы и могут быть при болезни лошади.  

In [25]:
outliers(horse_8col, 'pulse', 0)

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
3         1.0    9                39.1  164.0                         4.0   
41        2.0    9                39.0  150.0                         NaN   
55        1.0    9                38.6  160.0                         3.0   
255       1.0    9                38.8  184.0                         1.0   
275       1.0    9                38.8  150.0                         1.0   

     abdominal distension  packed cell volume  total protein  
3                     4.0                48.0            7.2  
41                    NaN                47.0            8.5  
55                    4.0                38.0            NaN  
255                   3.0                33.0            3.3  
275                   2.0                50.0            6.2

Частота сердечных сокращений при работе лошади
При работе пульс быстро возрастает, чтобы снабдить мышцы лошади кислородом и питательными веществами и удалить продукты метаболического распада. При полной нагрузке объём крови, перекачиваемой лошадиным сердцем, увеличивается в шесть раз! При подготовке лошади к пробегу основная работа проводится при пульсе не выше 160 уд/мин. Это считается аэробным уровнем, при котором мышцы наиболее рационально используют кислород. При пульсе свыше 160 уд/мин мышцы начинают работать анаэробно, неэффективно используя кислород и вырабатывая молочную кислоту. 

Источник: 50 хитростей и советов подготовки пробежной лошади. http://ksk-istok.ru/

Т.е. такие показатели пульса могут быть и у здоровой лошади при подготовке к соревнованиям, поэтому данные выбросы также являются реальными показателями.

In [27]:
outliers(horse_8col, 'packed cell volume', 0)

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
4         2.0    1                37.3  104.0                         NaN   
144       1.0    1                37.1   84.0                         3.0   
151       1.0    1                39.3    NaN                         4.0   

     abdominal distension  packed cell volume  total protein  
4                     NaN                74.0            7.4  
144                   4.0                75.0           81.0  
151                   2.0                75.0            NaN

Показатель packed cell volume - количество эритроцитов в крови по объему, нормальный диапазон от 30 до 50. Уровень повышается по мере циркуляции.
По данным сайта http://sibvetclinic.ru/ у здоровой лошади может быть до 60.

У лошади с заболеванием может быть выше, поэтому показатели также попадают в рамки допустимых.

In [19]:
outliers(horse_8col, 'total protein', 0)

Empty DataFrame
Columns: [surgery?, Age, rectal temperature, pulse, temperature of extremities, abdominal distension, packed cell volume, total protein]
Index: []

По показателю Total protein выбросов нет.

Помимо числовых столбцов можно проверить и категориальные столбцы. Из них выбросы заметны в столбце Age, т.к. максимальное значени в нем стоит 9, при допустимых 1 и 2, где:

          1 = взрослая лошадь
          2 = Молодой (<6 месяцев)

In [20]:
outliers(horse_8col, 'Age', 0)

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
3         1.0    9                39.1  164.0                         4.0   
9         2.0    9                38.3   90.0                         1.0   
13        2.0    9                38.0   92.0                         1.0   
16        1.0    9                 NaN  128.0                         3.0   
23        1.0    9                38.3  130.0                         NaN   
39        1.0    9                39.2  146.0                         NaN   
41        2.0    9                39.0  150.0                         NaN   
55        1.0    9                38.6  160.0                         3.0   
74        1.0    9                 NaN    NaN                         NaN   
75        1.0    9                39.7  100.0                         3.0   
82        1.0    9                38.1  100.0                         3.0   
103       1.0    9                38.0  140.0                         1.0   
109       1.0    9                38.4   84.0                         3.0   
135       1.0    9                38.1  136.0                         3.0   
191       2.0    9                 NaN  100.0                         2.0   
212       1.0    9                38.9  120.0                         1.0   
227       1.0    9                 NaN    NaN                         NaN   
229       1.0    9                38.5  120.0                         NaN   
244       1.0    9                38.2  124.0                         1.0   
252       1.0    9                38.8  124.0                         3.0   
255       1.0    9                38.8  184.0                         1.0   
262       2.0    9                38.6   88.0                         NaN   
275       1.0    9                38.8  150.0                         1.0   
287       2.0    9                39.5   84.0                         NaN   

     abdominal distension  packed cell volume  total protein  
3                     4.0                48.0            7.2  
9                     1.0                40.0            6.2  
13                    2.0                37.0            6.1  
16                    3.0                53.0            7.8  
23                    NaN                50.0           70.0  
39                    NaN                 NaN            NaN  
41                    NaN                47.0            8.5  
55                    4.0                38.0            NaN  
74                    NaN                37.0            4.9  
75                    NaN                48.0           57.0  
82                    1.0                36.0            5.7  
103                   2.0                39.0            5.3  
109                   3.0                36.0            6.6  
135                   3.0                33.0            4.9  
191                   1.0                37.0            4.7  
212                   3.0                47.0            6.3  
227                   NaN                45.0            6.5  
229                   NaN                35.0           54.0  
244                   4.0                47.0            8.0  
252                   4.0                50.0            7.6  
255                   3.0                33.0            3.3  
262                   NaN                35.0            5.9  
275                   2.0                50.0            6.2  
287                   NaN                28.0            5.0

In [66]:
# Определим какие значения и в каком соотношении стоят в столбце Age
qwe = horse_8col['Age'].value_counts().reset_index()
qwe['Age'] = qwe['Age'] / 300 * 100
qwe.columns = [ ['index', 'Age in percent'] ]
qwe

index Age in percent
0     1           92.0
1     9            8.0

92% значений столбца Age это "1" - взрослая лощадь, и остальные 8% значений это выбросы со значением "9", которго не может быть. Значения "2" - молодая лошадь вообще нет в таблице на текущий момент.
Заменим все "9" на "1".

In [63]:
# DataFrame с замененными значениями в столбце Age
re_horse_8col = horse_8col.copy()
re_horse_8col.loc[re_horse_8col['Age'] == 9, 'Age'] = 1
re_horse_8col

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
0         2.0    1                38.5   66.0                         3.0   
1         1.0    1                39.2   88.0                         NaN   
2         2.0    1                38.3   40.0                         1.0   
3         1.0    1                39.1  164.0                         4.0   
4         2.0    1                37.3  104.0                         NaN   
..        ...  ...                 ...    ...                         ...   
295       1.0    1                 NaN  120.0                         4.0   
296       2.0    1                37.2   72.0                         3.0   
297       1.0    1                37.5   72.0                         4.0   
298       1.0    1                36.5  100.0                         3.0   
299       1.0    1                37.2   40.0                         NaN   

     abdominal distension  packed cell volume  total protein  
0                     4.0                45.0            8.4  
1                     2.0                50.0           85.0  
2                     1.0                33.0            6.7  
3                     4.0                48.0            7.2  
4                     NaN                74.0            7.4  
..                    ...                 ...            ...  
295                   NaN                55.0           65.0  
296                   3.0                44.0            NaN  
297                   3.0                60.0            6.8  
298                   3.0                50.0            6.0  
299                   NaN                36.0           62.0  

[300 rows x 8 columns]

## Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [81]:
# Посчитаем количество пропусков
for col in re_horse_8col.columns:
    pct_missing_0 = re_horse_8col[col].isnull().sum()
    pct_missing_1 = re_horse_8col[col].isnull().mean()
    print(f'{col} - {pct_missing_1 :.2%}, Количество пропусков: {pct_missing_0}')

surgery? - 0.33%, Количество пропусков: 1
Age - 0.00%, Количество пропусков: 0
rectal temperature - 20.00%, Количество пропусков: 60
pulse - 8.00%, Количество пропусков: 24
temperature of extremities - 18.67%, Количество пропусков: 56
abdominal distension - 18.67%, Количество пропусков: 56
packed cell volume - 9.67%, Количество пропусков: 29
total protein - 11.00%, Количество пропусков: 33


In [262]:
# В столбце surgery? пропущено одно значение, удалим данную строку
sur_re_horse_8col = re_horse_8col.dropna(subset = ['surgery?'])

In [263]:
# заполним столбец 'rectal temperature' медианой, в зависимости от значения показателя 'surgery?' (наличия операции)
_sur_re_horse_8col = sur_re_horse_8col.copy()
_sur_re_horse_8col['rectal temperature'].fillna(sur_re_horse_8col.groupby('surgery?')['rectal temperature'].transform('median'), inplace = True)

In [264]:
# Столбец 'temperature of extremities'. Данный столбец должен коррелироваться с ректальной температурой. 
# Округлим значения ректальной температуры.
temp_ex_sur_re_horse_8col = _sur_re_horse_8col.copy()
temp_ex_sur_re_horse_8col['rectal temperature'] = temp_ex_sur_re_horse_8col['rectal temperature'].round()

temp_ex_sur_re_horse_8col['temperature of extremities'].fillna(temp_ex_sur_re_horse_8col.groupby('rectal temperature')['temperature of extremities'].transform('median'), inplace = True)
# Округлим значения в столбце до целого, т.к. медиана в некоторых ячейках оказалась не целым числом
temp_ex_sur_re_horse_8col['temperature of extremities'] = temp_ex_sur_re_horse_8col['temperature of extremities'].round()


In [265]:
# столбец pulse рассчитаем на основе surgery? и rectal temperature, т.к. у здоровой лошади, он не должен быть высоким, 
# также как и температура
pulse_ = temp_ex_sur_re_horse_8col.copy()

pulse_['pulse'].fillna(temp_ex_sur_re_horse_8col.groupby(['surgery?', 'rectal temperature'])['pulse'].transform('median'),inplace = True)

In [266]:
# добавил в общую таблицу вернозаполненные столбцы pulse и temperature of extremities
rty = _sur_re_horse_8col.join(pulse_, lsuffix = '_0', rsuffix = '_1')
rty = rty[['surgery?_0', 'Age_0', 'rectal temperature_0', 'pulse_1', 'temperature of extremities_1', 'abdominal distension_0', 'packed cell volume_0', 'total protein_0']]
rty.columns = ['surgery?','Age','rectal temperature','pulse','temperature of extremities','abdominal distension','packed cell volume','total protein']


In [267]:
# показатель abdominal distension' связан с хирургическим вмешательством (surgery). 
# Также определим его в зависимости от показателя temperature of extremities.
ab_dist = rty.copy()
ab_dist['abdominal distension'].fillna(rty.groupby(['surgery?','temperature of extremities'])['abdominal distension'].transform('median'), inplace = True)
ab_dist['abdominal distension'] = ab_dist['abdominal distension'].round()


In [268]:
# определим 'packed cell volume' от 'abdominal distension'
pcv = ab_dist.copy()
pcv['packed cell volume'].fillna(ab_dist.groupby('abdominal distension')['packed cell volume'].transform('median'), inplace = True)


In [261]:
#  'total protein' связан с обезвоживанием. Показатель, также как 'packed cell volume'. РАссчитаем в зависимости от него
tp = pcv.copy()
tp['total protein'].fillna(pcv.groupby('packed cell volume')['total protein'].transform('median'), inplace = True)
# медианой получилось заполнить не все пропуски. Осталось пустых 8 значений или 2.68%. 
# Заполним оставшиеся пропуски медианой по столбцу "total protein"
tp_ = tp.copy()
tp_['total protein'].fillna(tp['total protein'].median(), inplace = True)

In [269]:
for col in tp_.columns:
    pct_missing_0 = tp_[col].isnull().sum()
    pct_missing_1 = tp_[col].isnull().mean()
    print(f'{col} - {pct_missing_1 :.2%}, Количество пропусков: {pct_missing_0}')

surgery? - 0.00%, Количество пропусков: 0
Age - 0.00%, Количество пропусков: 0
rectal temperature - 0.00%, Количество пропусков: 0
pulse - 0.00%, Количество пропусков: 0
temperature of extremities - 0.00%, Количество пропусков: 0
abdominal distension - 0.00%, Количество пропусков: 0
packed cell volume - 0.00%, Количество пропусков: 0
total protein - 0.00%, Количество пропусков: 0


In [270]:
tp_

surgery?  Age  rectal temperature  pulse  temperature of extremities  \
0         2.0    1                38.5   66.0                         3.0   
1         1.0    1                39.2   88.0                         2.0   
2         2.0    1                38.3   40.0                         1.0   
3         1.0    1                39.1  164.0                         4.0   
4         2.0    1                37.3  104.0                         2.0   
..        ...  ...                 ...    ...                         ...   
295       1.0    1                38.1  120.0                         4.0   
296       2.0    1                37.2   72.0                         3.0   
297       1.0    1                37.5   72.0                         4.0   
298       1.0    1                36.5  100.0                         3.0   
299       1.0    1                37.2   40.0                         2.0   

     abdominal distension  packed cell volume  total protein  
0                     4.0                45.0            8.4  
1                     2.0                50.0           85.0  
2                     1.0                33.0            6.7  
3                     4.0                48.0            7.2  
4                     2.0                74.0            7.4  
..                    ...                 ...            ...  
295                   3.0                55.0           65.0  
296                   3.0                44.0            7.5  
297                   3.0                60.0            6.8  
298                   3.0                50.0            6.0  
299                   2.0                36.0           62.0  

[299 rows x 8 columns]